### PR TITLE
Pulsar producer implementation.

### PR DESCRIPTION
**Pull request by kmrdhruv**
TODO:: will add tests after first review.

### CODE REVIEW

File: pulsar/src/main/java/com/flipkart/varadhi/pulsar/entities/PulsarProducer.java [hunk 1]
```diff
@@ -2,19 +2,123 @@
...
+        messageBuilder.sendAsync().whenComplete((producerResult, throwable) -> {
+            if (producerResult != null) {
+                produceFuture.complete(new PulsarProducerResult(producerResult));
+            } else {
+                // In general Pulsar client and producer, auto-reconnects so this should be fine.Might need to
+                // refresh/re-create producer (and possibly client) if there are fatal errors, currently these
+                // failures are unknown.
+                produceFuture.completeExceptionally(throwable);
+            }
+        });
```
Comment (gauravAshok) [inline pulsar/src/main/java/com/flipkart/varadhi/pulsar/entities/PulsarProducer.java:53]:
```suggestion
        return messageBuilder.sendAsync().thenApply(r -> new PulsarProducerResult(r));
```

File: pulsar/src/main/java/com/flipkart/varadhi/pulsar/entities/PulsarProducer.java [hunk 2]
```diff
@@ -2,19 +2,123 @@
...
+        producerConfig.put("maxPendingMessages", maxPendingMessages);
+        // maxPendingMessages and maxPendingMessagesAcrossPartitions are kept same assuming worst case.
+        producerConfig.put("maxPendingMessagesAcrossPartitions", maxPendingMessages);
+        producerConfig.put("batchingMaxMessages", batchingMaxMessages);
+        producerConfig.put("batchingMaxBytes", batchingMaxMessages * getBatchingMaxBytes(batchingMaxMessages, topic));
+        return producerConfig;
     }
+
+    private String getProducerName(String topicName, String hostName) {
+        String topicSuffix = topicName.split("/")[4];
```
Comment (gauravAshok) [inline pulsar/src/main/java/com/flipkart/varadhi/pulsar/entities/PulsarProducer.java:105]: why split?

File: pulsar/src/main/java/com/flipkart/varadhi/pulsar/entities/PulsarProducer.java [hunk 3]
```diff
@@ -2,19 +2,123 @@
...
+        return String.format("%s.%s", topicSuffix, hostName);
+    }
+
+    private int getMaxPendingMessages(int topicMaxQps) {
+        // Assumption: Don't allow queue to build more than 1 second worth of messages.
+        // with bound [min, max]
+        // TODO:: This impacts memory so Discuss and close it.
+        return Math.min(MAX_PENDING_MESSAGES, Math.max(MIN_PENDING_MESSAGES, topicMaxQps));
```
Comment (gauravAshok) [inline pulsar/src/main/java/com/flipkart/varadhi/pulsar/entities/PulsarProducer.java:113]: I actually can;t comment right now. Anil did some work here. But here the even the topicQPS will be distributed across pods. So it is "technically" not correct to use topicMaxQps here.

I haven't experimented with produce configs like these, to comment. Dont have full understanding still.

File: pulsar/src/main/java/com/flipkart/varadhi/pulsar/entities/PulsarProducerResult.java [hunk 4]
```diff
@@ -1,11 +1,32 @@
...
+            throw new ArgumentException("Can not compare null ProducerResult.");
+        }
+        if (o instanceof PulsarProducerResult) {
+            return messageId.compareTo(((PulsarProducerResult) o).messageId);
+        } else {
+            throw new ArgumentException(String.format(
+                    "Can not compare different ProducerResult types. Expected ProducerResult is %s, given  %s.",
+                    PulsarProducerResult.class.getName(), o.getClass().getName()
+            ));
+        }
```
Comment (gauravAshok) [inline pulsar/src/main/java/com/flipkart/varadhi/pulsar/entities/PulsarProducerResult.java:26]: auto-generated?
Also, there is no requirement for ProduceResult to be comparable. Is there?

File: server/src/main/java/com/flipkart/varadhi/web/v1/produce/ProduceHandlers.java [hunk 5]
```diff
@@ -71,7 +71,8 @@ public List<RouteDefinition> get() {
     public void produce(RoutingContext ctx) {
         //TODO:: Request Validations pending
 
-        byte[] payload = ctx.body().asPojo(byte[].class);
+        byte[] payload = ctx.body().buffer().getBytes();
```
Comment (gauravAshok) [inline server/src/main/java/com/flipkart/varadhi/web/v1/produce/ProduceHandlers.java:74]: in other PR, you mentioned it creates a copy. skippable?

Comment (gauravAshok) [reply server/src/main/java/com/flipkart/varadhi/web/v1/produce/ProduceHandlers.java:74]: May be we can just use ctx.body().buffer().getByteBuf()?

File: common/src/main/java/com/flipkart/varadhi/exceptions/MessagingException.java [hunk 6]
```diff
@@ -0,0 +1,7 @@
+package com.flipkart.varadhi.exceptions;
+
+import lombok.experimental.StandardException;
+
+@StandardException
+public class MessagingException extends VaradhiException {
```
[the remark this answers is not in the document]
Comment (kmrdhruv) [reply common/src/main/java/com/flipkart/varadhi/exceptions/MessagingException.java:6]: base for any messaging tech specific exception. Added comment to the code.

File: entities/src/main/java/com/flipkart/varadhi/entities/CapacityPolicy.java [hunk 7]
```diff
@@ -1,5 +1,14 @@
...
+
+@Data
 public class CapacityPolicy {
-    String maxThroughput;
+    private int maxThroughputKBps;
+    private int maxQPS;
+
+    public CapacityPolicy(int maxQPS, int maxThroughputKBps) {
+        this.maxQPS = maxQPS;
+        this.maxThroughputKBps = maxThroughputKBps;
```
[the remark this answers is not in the document]
Comment (kmrdhruv) [reply entities/src/main/java/com/flipkart/varadhi/entities/CapacityPolicy.java:12]: retaining this as of now. We can revisit later, if its meaning need to change further.

File: messaging/src/main/java/com/flipkart/varadhi/produce/services/ProducerCache.java [hunk 8]
```diff
@@ -18,10 +18,22 @@ public ProducerCache(ProducerFactory producerFactory, String producerCacheBuilde
...
-    public Producer getProducer(StorageTopic storageTopic) throws ExecutionException {
-        return this.producerCache.get(
-                storageTopic.getName(),
-                () -> producerFactory.getProducer(storageTopic)
-        );
+    public Producer getProducer(StorageTopic storageTopic) {
+        try {
+            return this.producerCache.get(
+                    storageTopic.getName(),
+                    () -> producerFactory.getProducer(storageTopic)
```
[the remark this answers is not in the document]
Comment (kmrdhruv) [reply messaging/src/main/java/com/flipkart/varadhi/produce/services/ProducerCache.java:25]: correct. Caches are currently implemented as blocking call. These will be handled in unblocking manner in a separate PR.

File: messaging/src/main/java/com/flipkart/varadhi/produce/services/ProducerService.java [hunk 9]
```diff
@@ -97,4 +104,11 @@ private CompletableFuture<ProduceResult> produceToTopic(Message message, Interna
         return CompletableFuture.completedFuture(
                 ProduceResult.onNonProducingTopicState(messageId, internalTopic.getTopicState()));
     }
+
+    private String getPartitioningKey(Message message) {
+        if (message.hasHeader(GROUP_ID)) {
+            return message.getHeader(GROUP_ID);
+        }
+        return stringGenerator.generate(RANDOM_PARTITION_KEY_LENGTH);
```
[the remark this answers is not in the document]
Comment (kmrdhruv) [reply messaging/src/main/java/com/flipkart/varadhi/produce/services/ProducerService.java:112]: will move this to producer as per the discussion.

File: pulsar/src/main/java/com/flipkart/varadhi/pulsar/PulsarProducerFactory.java [hunk 10]
```diff
@@ -1,13 +1,37 @@
...
+        this.clientProvider = clientProvider;
+        this.hostName = hostName;
+        this.producerOptions = null == producerOptions ? new ProducerOptions() : producerOptions;
+    }
+
     @Override
     public Producer getProducer(PulsarStorageTopic storageTopic) {
-        return new PulsarProducer();
+        try {
+            return new PulsarProducer(clientProvider, storageTopic, producerOptions, hostName);
```
[the remark this answers is not in the document]
Comment (kmrdhruv) [reply pulsar/src/main/java/com/flipkart/varadhi/pulsar/PulsarProducerFactory.java:30]: correct. All caches will be converted to non blocking in a separate PR.

File: pulsar/src/main/java/com/flipkart/varadhi/pulsar/entities/PulsarProducer.java (hunk 3 above)
Comment (kmrdhruv) [reply pulsar/src/main/java/com/flipkart/varadhi/pulsar/entities/PulsarProducer.java:113]: sure. This needs to be tuned based on benchmark and observations. As of now approx calculation (assuming worst case of all messages to the same producer) to allow messages worth 1 sec with both lower/upper bound.

File: entities/src/testFixtures/java/com/flipkart/varadhi/entities/CapacityHelper.java [hunk 11]
```diff
@@ -0,0 +1,7 @@
+package com.flipkart.varadhi.entities;
+
+public class CapacityHelper {
+    public static CapacityPolicy getDefault() {
```
[the remark this answers is not in the document]
Comment (kmrdhruv) [reply entities/src/testFixtures/java/com/flipkart/varadhi/entities/CapacityHelper.java:4]: Though this was for testing purpose, but I think it should be fine to move to CapacityPolicy as well.

File: common/build.gradle [hunk 12]
```diff
@@ -3,7 +3,6 @@ plugins {
 }
 
 dependencies {
-    implementation("io.vertx:vertx-core")
```
[the remark this answers is not in the document]
Comment (kmrdhruv) [reply common/build.gradle:6]: it was not being used.
As discussed, it should be fine to keep Vertx core internal to Varadhi, but not in externally visible contract.

File: pulsar/src/main/java/com/flipkart/varadhi/pulsar/entities/PulsarProducer.java (hunk 2 above)
Comment (kmrdhruv) [reply pulsar/src/main/java/com/flipkart/varadhi/pulsar/entities/PulsarProducer.java:105]: can be left as is for now. We will evaluate if we need to shorten it later.

File: pulsar/src/main/java/com/flipkart/varadhi/pulsar/entities/PulsarProducer.java [hunk 13]
```diff
@@ -2,19 +2,123 @@
...
+import org.apache.pulsar.client.api.TypedMessageBuilder;
 
+import java.util.Map;
 import java.util.concurrent.CompletableFuture;
 
+import static com.flipkart.varadhi.pulsar.Constants.Producer.*;
 
 @Slf4j
 public class PulsarProducer implements Producer {
+
```
[the remark this answers is not in the document]
Comment (kmrdhruv) [reply pulsar/src/main/java/com/flipkart/varadhi/pulsar/entities/PulsarProducer.java:20]: We can move it as understanding evolves.

File: pulsar/src/main/java/com/flipkart/varadhi/pulsar/entities/PulsarProducer.java [hunk 14]
```diff
@@ -2,19 +2,123 @@
...
-        result.complete(new PulsarProducerResult());
-        return result;
+    public CompletableFuture<ProducerResult> ProduceAsync(String partitioningKey, Message message) {
+
+        CompletableFuture<ProducerResult> produceFuture = new CompletableFuture<>();
+
+        TypedMessageBuilder<byte[]> messageBuilder =
+                pulsarProducer.newMessage().key(partitioningKey).value(message.getPayload());
+
+        message.getRequestHeaders().asMap()
```
[the remark this answers is not in the document]
Comment (kmrdhruv) [reply pulsar/src/main/java/com/flipkart/varadhi/pulsar/entities/PulsarProducer.java:41]: intent is to have multi value transferred to the consumption path as well. Pulsar was overriding with last value. Modified it to have single value when persisting in Pulsar, which combines all the member values from the collection.

File: server/src/main/java/com/flipkart/varadhi/web/v1/produce/ProduceHandlers.java (hunk 5 above)
Comment (kmrdhruv) [reply server/src/main/java/com/flipkart/varadhi/web/v1/produce/ProduceHandlers.java:74]: As of now continuing with ctx.body().buffer().getBytes() (which involves extra copy). Added a TODO to look this into independent of current PR.
getByteBuf() gives access to backing array, but would require Varadhi to maintain (or consume based on) start/end offset of the messages, need more thoughts.

File: pulsar/src/main/java/com/flipkart/varadhi/pulsar/entities/PulsarProducerResult.java (hunk 4 above)
Comment (kmrdhruv) [reply pulsar/src/main/java/com/flipkart/varadhi/pulsar/entities/PulsarProducerResult.java:26]: U mean ProduceResult or ProducerResult ?
For former comparable is not required as it is more then message offset.  Renamed this class to be Offset and PulsarOffset, however retained the Comparable for Offset.

**Pull request merged**